File: src/main/java/com/ort/app/web/controller/NewVillageController.java
package com.ort.app.web.controller;

import com.ort.app.logic.VillageParticipateLogic;
import com.ort.app.web.controller.assist.NewVillageAssist;
import com.ort.app.web.exception.WolfMansionBusinessException;
import com.ort.app.web.form.NewVillageForm;
import com.ort.app.web.form.validator.NewVillageFormValidator;
import com.ort.dbflute.exbhv.CharaBhv;
import com.ort.dbflute.exbhv.CharaGroupBhv;
import com.ort.dbflute.exentity.Chara;
import com.ort.dbflute.exentity.CharaGroup;
import com.ort.dbflute.exentity.Village;
import com.ort.fw.security.UserInfo;
import com.ort.fw.util.WolfMansionUserInfoUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;

@Controller
public class NewVillageController {

    // ===================================================================================
    //                                                                           Attribute
    //                                                                           =========
    @Autowired
    private CharaGroupBhv charaGroupBhv;
    @Autowired
    private CharaBhv charaBhv;
    @Autowired
    private NewVillageFormValidator newVillageFormValidator;
    @Autowired
    private NewVillageAssist newVillageAssist;
    @Autowired
    private VillageParticipateLogic participateLogic;

    @InitBinder("villageForm")
    public void initBinder(WebDataBinder binder) {
        binder.addValidators(newVillageFormValidator);
    }

    // ===================================================================================
    //                                                                             Execute
    //                                                                             =======
    // 新規村作成初期表示
    @RequestMapping("/new-village") // 戻る場合もあるのでpostもok
    private String newVillageIndex(NewVillageForm form, Model model) {
        newVillageAssist.setIndexModel(form, model);
        return "new-village";
    }

    // 新規村作成_流用
    @PostMapping("/new-village/divert/{villageId}")
    private String divert(@PathVariable Integer villageId, NewVillageForm form, Model model) {
        newVillageAssist.setIndexModel(form, model);
        newVillageAssist.override(form, model, villageId);
        return "new-village";
    }

    // 新規村作成確認画面
    @PostMapping("/new-village/confirm")
    private String newVillageConfirm(@Validated @ModelAttribute("villageForm") NewVillageForm villageForm, BindingResult bindingResult,
                                     Model model) {
        if (bindingResult.hasErrors()) {
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        // ログインしていなかったらNG
        UserInfo userInfo = WolfMansionUserInfoUtil.getUserInfo();
        if (userInfo == null) {
            model.addAttribute("errorMessage", "ログインし直してください。");
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        // 決着のついていない参加していたり建てた村が終了していなかったらNG
        if (participateLogic.isParticipatingOrCreatingVillage()) {
            model.addAttribute("errorMessage", "参加/村建てした村の決着がつくまでは村を建てられません。");
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        // キャラチップ数が足りなかったらNG
        if (!newVillageAssist.isEnoughCharacterNum(villageForm)) {
            model.addAttribute("errorMessage", "定員に対してキャラ数が不足しています。");
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }

        // キャラセット名を取得
        CharaGroup charaGroup =
                charaGroupBhv.selectEntityWithDeletedCheck(cb -> cb.query().setCharaGroupId_Equal(villageForm.getCharacterSetId()));
        String charaGroupName = charaGroup.getCharaGroupName();
        model.addAttribute("characterSetName", charaGroupName);
        // キャラ画像URL
        Chara chara = charaBhv.selectEntityWithDeletedCheck(cb -> {
            cb.query().setCharaId_Equal(villageForm.getDummyCharaId());
            cb.query().setCharaGroupId_Equal(charaGroup.getCharaGroupId());
        });
        charaBhv.loadCharaImage(chara, charaImageCB -> {
            charaImageCB.query().setFaceTypeCode_Equal_通常();
        });
        model.addAttribute("dummyCharaName", chara.getCharaName());
        model.addAttribute("characterImgUrl", chara.getNormalCharaImgUrl());
        model.addAttribute("characterImgWidth", chara.getDisplayWidth());
        model.addAttribute("characterImgHeight", chara.getDisplayHeight());

        // 時間と日時の表示
        model.addAttribute("startDateTime", makeStartDateTime(villageForm));
        model.addAttribute("interval", makeInterval(villageForm));

        return "new-village-confirm";
    }

    // 新規村作成
    @PostMapping("/new-village/create")
    private String makeVillage(
            @Validated @ModelAttribute("villageForm") NewVillageForm villageForm,
            BindingResult bindingResult,
            Model model
    ) {
        if (bindingResult.hasErrors()) {
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        // ログインしていなかったらNG
        UserInfo userInfo = WolfMansionUserInfoUtil.getUserInfo();
        if (userInfo == null) {
            model.addAttribute("errorMessage", "ログインし直してください。");
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        // 決着のついていない参加していたり建てた村が終了していなかったらNG
        if (participateLogic.isParticipatingOrCreatingVillage()) {
            model.addAttribute("errorMessage", "参加/村建てした村の決着がつくまでは村を建てられません。");
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        Village village = null;
        try {
            village = newVillageAssist.createVillage(villageForm, userInfo.getUsername());
        } catch (WolfMansionBusinessException e) {
            model.addAttribute("errorMessage", e.getMessage());
            newVillageAssist.setIndexModel(villageForm, model);
            return "new-village";
        }
        return "redirect:/village/" + village.getVillageId() + "#bottom";
    }

    // ===================================================================================
    //                                                                        Assist Logic
    //                                                                        ============
    private String makeInterval(NewVillageForm villageForm) {
        String hour = String.format("%02d時間", villageForm.getDayChangeIntervalHours());
        String minute = String.format("%02d分", villageForm.getDayChangeIntervalMinutes());
        String second = String.format("%02d秒", villageForm.getDayChangeIntervalSeconds());
        return hour + minute + second;
    }

    private LocalDateTime makeStartDateTime(NewVillageForm villageForm) {
        return LocalDateTime.of(villageForm.getStartYear(), villageForm.getStartMonth(), villageForm.getStartDay(),
                villageForm.getStartHour(), villageForm.getStartMinute(), 0);
    }
}

File: src/main/kotlin/com/ort/app/api/NewVillageController.kt
package com.ort.app.api

import com.ort.app.api.request.NewVillageForm
import com.ort.app.api.request.validator.NewVillageFormValidator
import com.ort.app.api.view.NewVillageDivertContent
import com.ort.app.api.view.OptionContent
import com.ort.app.application.coordinator.VillageCoordinator
import com.ort.app.application.service.CharaService
import com.ort.app.application.service.PlayerService
import com.ort.app.application.service.VillageService
import com.ort.app.domain.model.skill.Skill
import com.ort.app.domain.model.village.VillageStatus
import com.ort.app.fw.exception.WolfMansionBusinessException
import com.ort.app.fw.util.WolfMansionUserInfoUtil
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.InitBinder
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

@Controller
class NewVillageController(
    private val newVillageFormValidator: NewVillageFormValidator,
    private val villageCoordinator: VillageCoordinator,
    private val villageService: VillageService,
    private val charaService: CharaService,
    private val playerService: PlayerService
) {

    @InitBinder("villageForm")
    fun initBinder(binder: WebDataBinder) {
        binder.addValidators(newVillageFormValidator)
    }

    // 新規村作成初期表示
    @RequestMapping("/new-village") // 戻る場合もあるのでpostもok
    private fun newVillageIndex(form: NewVillageForm, model: Model): String {
        setIndexModel(form, model)
        return "new-village"
    }

    // 新規村作成_流用
    @PostMapping("/new-village/divert/{villageId}")
    private fun divert(
        @PathVariable villageId: Int,
        form: NewVillageForm,
        model: Model
    ): String {
        setIndexModel(form, model)
        val village = villageService.findVillage(villageId)
            ?: throw IllegalStateException("village not found.")
        form.override(village)
        model.addAttribute("villageForm", form)
        return "new-village"
    }

    // 新規村作成確認画面
    @PostMapping("/new-village/confirm")
    private fun newVillageConfirm(
        @Validated @ModelAttribute("villageForm") villageForm: NewVillageForm,
        result: BindingResult,
        model: Model
    ): String {
        val player = WolfMansionUserInfoUtil.getUserInfo()?.let {
            playerService.findPlayer(it.username)
        }
        val charachip = charaService.findCharachip(villageForm.characterSetId!!)
        if (result.hasErrors() || player == null || charachip == null) {
            setIndexModel(villageForm, model)
            return "new-village"
        }
        try {
            villageCoordinator.assertCreateVillage(player, villageForm.personMaxNum!!, charachip)
        } catch (e: WolfMansionBusinessException) {
            model.addAttribute("errorMessage", e.message)
            setIndexModel(villageForm, model)
            return "new-village"
        }
        model.addAttribute("characterSetName", charachip.name)
        val dummyChara = charachip.charas.chara(villageForm.dummyCharaId!!)
        model.addAttribute("dummyCharaName", dummyChara.name)
        model.addAttribute("characterImgUrl", dummyChara.defaultImage().url)
        model.addAttribute("characterImgWidth", dummyChara.size.width)
        model.addAttribute("characterImgHeight", dummyChara.size.height)

        // 時間と日時の表示
        model.addAttribute("startDateTime", mapStartDateTime(villageForm))
        model.addAttribute("interval", mapInterval(villageForm))
        return "new-village-confirm"
    }

    // 新規村作成
    @PostMapping("/new-village/create")
    private fun createVillage(
        @Validated @ModelAttribute("villageForm") villageForm: NewVillageForm,
        result: BindingResult,
        model: Model
    ): String {
        val player = WolfMansionUserInfoUtil.getUserInfo()?.let {
            playerService.findPlayer(it.username)
        }
        val charachip = charaService.findCharachip(villageForm.characterSetId!!)
        if (result.hasErrors() || player == null || charachip == null) {
            setIndexModel(villageForm, model)
            return "new-village"
        }
        val village = try {
            villageCoordinator.assertCreateVillage(player, villageForm.personMaxNum!!, charachip)
            villageCoordinator.registerVillage(villageForm.toVillage(player), villageForm.dummyJoinMessage!!)
        } catch (e: WolfMansionBusinessException) {
            model.addAttribute("errorMessage", e.message)
            setIndexModel(villageForm, model)
            return "new-village"
        }
        return "redirect:/village/${village.id}#bottom"
    }

    // ===================================================================================
    //                                                                        Assist Logic
    //                                                                        ============
    private fun mapInterval(villageForm: NewVillageForm): String {
        val hour = String.format("%02d時間", villageForm.dayChangeIntervalHours)
        val minute = String.format("%02d分", villageForm.dayChangeIntervalMinutes)
        val second = String.format("%02d秒", villageForm.dayChangeIntervalSeconds)
        return hour + minute + second
    }

    private fun mapStartDateTime(villageForm: NewVillageForm): String {
        return LocalDateTime.of(
            villageForm.startYear!!, villageForm.startMonth!!, villageForm.startDay!!,
            villageForm.startHour!!, villageForm.startMinute!!, 0
        ).format(DateTimeFormatter.ofPattern("uuuu/MM/dd HH:mm"))
    }

    private fun setIndexModel(form: NewVillageForm, model: Model) {
        form.initialize()
        model.addAttribute("villageForm", form)

        val finishedVillages = villageService.findVillages(
            (VillageStatus.settledStatusList + VillageStatus.finishedStatusList).distinct().map { VillageStatus(it) }
        ).list.map { NewVillageDivertContent(it.id, it.id.toString().padStart(4, '0'), it.name) }
        model.addAttribute("villageList", finishedVillages)
        model.addAttribute("skillListStr", Skill.getSkillListStr())
        model.addAttribute("nowYear", LocalDateTime.now().year)
        val charachips = charaService.findCharachips().list.map {
            OptionContent(name = "${it.name}（${it.designer.name}様作）", value = it.id.toString())
        }
        model.addAttribute("characterSetList", charachips)
    }
}